### PR TITLE
Add a KubeVirt CR option to use a dedicated migration network

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This PR adds the `dedicatedMigrationNetwork` migration option, to specify a multus network to use for migrations.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This requires a multus network already configured to link all the nodes together and provide a DHCP server.

**Release note**:

```release-note
Migrations can now be done over a dedicated multus network
```

### CODE REVIEW

File: pkg/virt-handler/vm.go
Comment (rmohr) [reply pkg/virt-handler/vm.go:708]: Would it make sense to use the pod [downward api](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api) to get `k8s.v1.cni.cncf.io/networks-status` annotation info into virt-handler and in addition a commandline flag for virt-handler where you pass in the multus network name?

One can then look up info like this:

```
    k8s.v1.cni.cncf.io/networks-status: |- 
      [{
          "name": "openshift-sdn",
          "interface": "eth0",
          "ips": [
              "10.128.2.14"
          ],
          "default": true,
          "dns": {}
      },{
          "name": "macvlan-bridge",
          "interface": "net1",
          "ips": [
              "20.2.2.100"
          ],
          "mac": "22:2f:60:a5:f8:00",
          "dns": {}
      }]
```

This would probably remove all guessing and you could simply read the IP from this info, since you would know the multus network name from the commandline flag.

Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: Force-pushed switch to downward API, let me know what you think!
Haven't done the commandline part yet. Would it make sense to parse the json part of the NetworkAttachmentDefinition to find the network name and use that as a key instead of the interface name?

File: pkg/virt-handler/vm.go
Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: Not sure what you mean there.. The main pod IP address is actually just passed as a commandline argument.

File: pkg/virt-handler/vm.go
Comment (rmohr) [reply pkg/virt-handler/vm.go]: Yes, I think so. Just using the multus name as key and then extracting all info from there sounds great.

File: pkg/virt-handler/vm.go
Comment (rmohr) [reply pkg/virt-handler/vm.go]: Ah I see, I misinterpreted what this does. Will add more precise suggestions.

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:700]: Can you move this whole `findMigrationIP` to the virt-handler startup phase and just set d.ipAddress upfront on the controller? It can not change during the handler lifetime and it shoud keep the core-logic more in the focus then.

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:705]: Let's see that we return errors. I would suggest that you detect that we are supposed to use a multus network during the startup phase and only then call this function. Then if you see an error, fail the whole handler start.

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:712]: As suggested by you in another comment, here it would be great if you would use the multus network name and extract all necessary info from there. Only this way it does not matter if we have one or more multus netwoks. Otherwise any kind of preset mechanism which would add multus annotations could cause trouble.

File: pkg/virt-handler/vm.go
Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: So, I started working on that, using https://github.com/k8snetworkplumbingwg/network-attachment-definition-client to access the NetworkAttachmentDefinition.
However, the client above fails to compile. I'm getting obscure errors I wouldn't know how to fix. (Starts with: `vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/typed/k8s.cni.cncf.io/v1/k8s.cni.cncf.io_client.go:74:32: undefined: serializer.DirectCodecFactory`).
Even if it worked, finding the NetworkAttachmentDefinition and parsing the json blob it contains adds a lot of complxity to the code, so I'm not sure it's worth it...
Maybe we could request that the name of the network matches the name of the NetworkAttachmentDefinition? Or add fields in the CR? (we'll need a second one anyway if we want to allow using NetworkAttachmentDefinitions that are in a different namespace than `kubevirt`)
@rmohr please let me know what you think! Thanks

Comment (rmohr) [reply pkg/virt-handler/vm.go]: @jean-edouard did you consider simply parsing the json into a `map[string]interface{}`? But in general you don't need the client. You only need the api definition to deserialize the json. Consider simply copying the `NetworkStatus` into our codebase (https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/master/pkg/apis/k8s.cni.cncf.io/v1/types.go#L101). It is nice if we can reuse external code, but at the end there is a `v1` REST API defined by it, which means that we can define our own struct for the API without risks.

Comment (rmohr) [reply pkg/virt-handler/vm.go]: But looking at your latest code, you basically already did that. So maybe I did not understand what you try to do? I think in general you will not have to ever look up a `NetworkAttachmentDefinition` in the cluster if I did not miss anything.

Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: This is probably a lack of knowledge on my side, but is there a way to retrieve a NetworkAttachmentDefinition from k8s in golang without having to use a client like network-attachment-definition-client?
If so, could you please point me to some docs/example on how to do that?
Thanks!

Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: Ah, so maybe this is a misunderstanding. In the CR, the cluster admin provides a NetworkAttachmentDefinition by name.
Inside that object there is a json blob (don't ask me why), with a different name, and that's the name I need to parse the yaml network info!

Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: This was discussed offline, and Roman may have found a specific notation that could potentially alleviate this hole issue.
I'm looking into that now.

Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: Indeed, the network annotation can name the interface using the "@" notation. Pushed, thanks!

**Comment by jean-edouard** [comment]
Depends on https://github.com/kubevirt/kubevirtci/pull/671

**Comment by jean-edouard** [comment]
/hold
There's a couple things wrong with this PR, no need to waste resources testing/reviewing it, will unhold when fixed.

**Comment by jean-edouard** [comment]
/hold cancel
- Functest moved to network module to get CNAO and whereabouts
- Last PR comments addressed in new commit
- Rebased

**Comment by jean-edouard** [comment]
/hold cancel
Assumed all along that sig-network requested at least 1 secondary network... Added.
Also moved to more harmless IP range. Sorry for all the noise!

File: pkg/virt-config/config-map.go
Comment (rmohr) [inline pkg/virt-config/config-map.go]: I think you can remove this. We don't add new options to the config-map anymore. If people would try to set it, we would just ignore it and stick with the default.

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:702]: I think we were going forth and back a little bit here on using the downwards api or directly reading the IP from the interface. What was the final reason for using the downward api (which is fine by me in general)?

File: pkg/virt-handler/vm.go
Comment (rmohr) [inline pkg/virt-handler/vm.go:1464]: Would it make sense to look the ip address up outside of the controller and pass it as the desired IP address instead of implicitly overriding the passed in IP address ad creation time? The rename to `migrationIPAddress` is great if that is really the only reason why we ever use it.

**Review by rmohr — COMMENTED** [review]
A first pass over the code.

File: pkg/virt-config/config-map.go
Comment (jean-edouard) [reply pkg/virt-config/config-map.go]: I wish! The project won't compile if this struct is not identical to the CR struct.

File: pkg/virt-handler/vm.go
Comment (jean-edouard) [reply pkg/virt-handler/vm.go:702]: I'm pretty sure that was an offline suggestion from you, and it does look cleaner using the downwards API. Not sure if there's much more to it.

Comment (rmohr) [reply pkg/virt-handler/vm.go:702]: Ack.

File: pkg/virt-config/config-map.go
Comment (rmohr) [reply pkg/virt-config/config-map.go]: Ack. With https://github.com/kubevirt/kubevirt/pull/6784 it will not have a relevance anymore.

File: pkg/virt-handler/vm.go
Comment (jean-edouard) [reply pkg/virt-handler/vm.go]: Done!

File: automation/test.sh
Comment (rmohr) [inline automation/test.sh:46]: This is the default anyway, right?

File: cmd/virt-handler/virt-handler.go
Comment (rmohr) [inline cmd/virt-handler/virt-handler.go:351]: Maybe let's avoid the pointer magic. How about something like this:


```
findMigrationIP(defaultMigrationIP string) (string, error) {}
```

and then 


```
migrationIPAddress, err := findMigrationIP(app.PodIPAddress)
```

File: cmd/virt-handler/virt-handler.go
Comment (rmohr) [inline cmd/virt-handler/virt-handler.go:190]: Do we rely on some implicit behaviour that we can deal with empty files? Because it looks like we would crate this file always and in case that there is no annotation the file is presumably empty?

**Review by rmohr — CHANGES_REQUESTED** [review]

File: automation/test.sh
Comment (jean-edouard) [reply automation/test.sh:46]: No! I was surprised too

File: cmd/virt-handler/virt-handler.go
Comment (jean-edouard) [reply cmd/virt-handler/virt-handler.go:190]: The file will always contain at least 1 entry for the main network.
Also yes, `yaml.Unmarshall()` will happily decode an empty string into an empty list.

Comment (rmohr) [reply cmd/virt-handler/virt-handler.go:190]: Even if multus is not installed? The annotation looks very multus specific.

Comment (jean-edouard) [reply cmd/virt-handler/virt-handler.go:190]: The annotation is multus-specific, but the network status isn't.
If the annotation is not present and/or multus is not installed, the network-status file will still show the status of the main network.
Or maybe I misunderstood your question? :)

Comment (rmohr) [reply cmd/virt-handler/virt-handler.go:190]: Well maybe I misunderstand something, but 

```
	container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
		Name:      "podinfo",
		MountPath: "/etc/podinfo",
	})
	pod.Volumes = append(pod.Volumes, corev1.Volume{
		Name: "podinfo",
		VolumeSource: corev1.VolumeSource{
			DownwardAPI: &corev1.DownwardAPIVolumeSource{
				Items: []corev1.DownwardAPIVolumeFile{
					{
						Path: "network-status",
						FieldRef: &corev1.ObjectFieldSelector{
							FieldPath: `metadata.annotations['k8s.v1.cni.cncf.io/network-status']`,
						},
					},
				},
			},
		},
	})
	```
	
seems to be the source of the `network-status` file and that one looks to me like it would only contain any context if this annotation exists. And this annotation does probably not exist if there is not multus installed?

Comment (jean-edouard) [reply cmd/virt-handler/virt-handler.go:190]: Sorry, you are absolutely right!
Without multus, /etc/podinfo/network-status still gets created but is empty.
As I said before though, this case is properly handled by the code, though not explicitly.

Comment (jean-edouard) [reply cmd/virt-handler/virt-handler.go:193]: Added unit tests

File: cmd/virt-handler/virt-handler_test.go
Comment (rmohr) [inline cmd/virt-handler/virt-handler_test.go:62]: I think we should clean up the temporary files. They pile up in the tmp directory otherwise.

Comment (jean-edouard) [reply cmd/virt-handler/virt-handler_test.go:62]: There's a deferred removal below, do we need something else?

**Comment by jean-edouard** [comment]
> @jean-edouard @rmohr what needs to happen to converge this?

Sorry this is taking so long...
I just replied to a comment that I previously missed.
The PR also needs a rebase which I'll take care of right now.

File: cmd/virt-handler/virt-handler_test.go
Comment (rmohr) [reply cmd/virt-handler/virt-handler_test.go:62]: great then :)

File: cmd/virt-handler/virt-handler.go
Comment (rmohr) [reply cmd/virt-handler/virt-handler.go:193]: Thanks! :)

**Comment by rmohr** [comment]
/retest
/approve

@xpivarc do you want to review the PR a last time and lgtm?

File: tests/network/vmi_networking.go
Comment (xpivarc) [reply tests/network/vmi_networking.go:892]: I think it could be done through conditions on the `kv` but I am also fine with your approach. I would just ask that we move the `Eventually` block into `SetDedicatedMigrationNetwork` if we ever going to use it again.

File: pkg/virt-operator/resource/generate/components/deployments.go
Comment (xpivarc) [inline pkg/virt-operator/resource/generate/components/deployments.go:141]: I would move this to `NewHandlerDaemonSet` as it is only needed there.

**Review by xpivarc — APPROVED** [review]
Great work! 
My only real concern is the `AfterEach` block.
/hold

**Comment by jean-edouard** [comment]
Thank you @xpivarc !
I agree with all your comments and have addressed them in a new commit.
PTAL!

File: tests/migration.go
Comment (xpivarc) [inline tests/migration.go:110]: This is not a bug. `UpdateKubeVirtConfigValueAndWait` is only waiting for the new config to be propagated but that doesn't guarantee that operator had time to do the job.

**Review by xpivarc — APPROVED** [review]
Thank you!
/hold cancel 
Ping me if you decide to adjust the comment.

File: tests/migration.go
Comment (xpivarc) [inline tests/migration.go:121]: I think we need to check the new handlers are ready otherwise we will see this failure https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6036/pull-kubevirt-e2e-k8s-1.20-sig-network/1468980795975667712

**Review by xpivarc — COMMENTED** [review]
/hold

**Comment by stu-gott** [comment]
@xpivarc does this still need to have a hold label?

File: tests/migration.go
Comment (xpivarc) [inline tests/migration.go:123]: We are only checking they are running not ready. It would be better to check for readiness but we don't advertise the device plugin readiness.

**Review by xpivarc — APPROVED** [review]
/hold cancel

**Comment by jean-edouard** [comment]
Moved tested virt-handler code from cmd to pkg and rebased

**Review by xpivarc — APPROVED** [review]
@jean-edouard Thank you!

**Comment by rmohr** [comment]
Finally 😎😁

**Pull request merged**
